AP_DDS: Parameters service

PQuill33 opened the pull request:
### ISSUE ###
[#28292 ](https://github.com/ArduPilot/ardupilot/issues/28292)

### Changes ###
-Add set parameters service call to AP_DDS library
- service is /ap/set_parameters
- Add get parameters service call to AP_DDS library
- service is /ap/get_parameters

### Test ### 
Using ros2 CLI, build and launch ardupilot sitl
```
colcon build --packages-up-to ardupilot_sitl
source install/setup.bash
ros2 launch ardupilot_sitl sitl_dds_udp.launch.py transport:=udp4 synthetic_clock:=True wipe:=False model:=quad speedup:=1 slave:=0 instance:=0 defaults:=$(ros2 pkg prefix ardupilot_sitl)/share/ardupilot_sitl/config/default_params/copter.parm,$(ros2 pkg prefix ardupilot_sitl)/share/ardupilot_sitl/config/default_params/dds_udp.parm sim_address:=127.0.0.1 master:=tcp:127.0.0.1:5760 sitl:=127.0.0.1:5501
```

- Check that parameter service is up
```
ros2 service list
```
Expected output
![service_list](https://github.com/user-attachments/assets/9a59f4f3-91f8-4233-a6da-514444ad466e)

- Service call to set parameter
```
ros2 service call /ap/set_parameters rcl_interfaces/srv/SetParameters "{parameters: [{name: LOIT_SPEED, value: {type: 2, integer_value: 1250}}]}"
```
Expected output
mavproxy window: 
![param_success](https://github.com/user-attachments/assets/1459d43e-30e2-41a1-b875-69745b5bc3bb)


service call window:
![single_param_success](https://github.com/user-attachments/assets/0db6e55e-87dc-4acc-a49e-6d6fc42e52d7)

- Service call to get parameter
```
ros2 service call /ap/get_parameters rcl_interfaces/srv/GetParameters "{names: ['LOIT_SPEED']}"
```
Expected output
mavproxy window:
![get_param_success](https://github.com/user-attachments/assets/13cf50ae-2067-4b8f-89c6-6525ae8c0e88)

service call window:
![get_param_response](https://github.com/user-attachments/assets/cf17944c-1a1e-4311-8e93-319d250ba9ae)

tizianofiorenzani commented:
I have tested with the Plane SITL, trying to ask for a wrong parameter. I mispelled `AIRSPEED_CRUISE`, which should be `22.0`, and I expected to get some error. The only thing that tells me the response is wrong is the `type = 0`.
```
ros2 service call /ap/get_parameters rcl_interfaces/srv/GetParameters "names: [AIRSPEED_CRUIS2E]"
waiting for service to become available...
requester: making request: rcl_interfaces.srv.GetParameters_Request(names=['AIRSPEED_CRUIS2E'])

response:
rcl_interfaces.srv.GetParameters_Response(values=[rcl_interfaces.msg.ParameterValue(type=0, bool_value=True, integer_value=19513347822207839, double_value=22.0, string_value='', byte_array_value=[], bool_array_value=[], integer_array_value=[], double_array_value=[], string_array_value=[])])

```

PQuill33 commented:
> I have tested with the Plane SITL, trying to ask for a wrong parameter. I mispelled `AIRSPEED_CRUISE`, which should be `22.0`, and I expected to get some error. The only thing that tells me the response is wrong is the `type = 0`.
> 
> ```
> ros2 service call /ap/get_parameters rcl_interfaces/srv/GetParameters "names: [AIRSPEED_CRUIS2E]"
> waiting for service to become available...
> requester: making request: rcl_interfaces.srv.GetParameters_Request(names=['AIRSPEED_CRUIS2E'])
> 
> response:
> rcl_interfaces.srv.GetParameters_Response(values=[rcl_interfaces.msg.ParameterValue(type=0, bool_value=True, integer_value=19513347822207839, double_value=22.0, string_value='', byte_array_value=[], bool_array_value=[], integer_array_value=[], double_array_value=[], string_array_value=[])])
> ```
> 
> Instead, if you ask for the correct parameter I have:
> 
> ```
> ros2 service call /ap/get_parameters rcl_interfaces/srv/GetParameters "names: [AIRSPEED_CRUISE]"
> waiting for service to become available...
> requester: making request: rcl_interfaces.srv.GetParameters_Request(names=['AIRSPEED_CRUISE'])
> 
> response:
> rcl_interfaces.srv.GetParameters_Response(values=[rcl_interfaces.msg.ParameterValue(type=3, bool_value=True, integer_value=19513347822207839, double_value=24.0, string_value='', byte_array_value=[], bool_array_value=[], integer_array_value=[], double_array_value=[], string_array_value=[])])
> ```
> 
> Overall seems correct. I have tried to set and get multiple parameters, changed the airspeed in flight, made sure it was effective.

The error gets reported as a GCS message. It should read `AP: DDS: Get Parameters Request : ONE OR MORE PARAM NOT FOUND`. 

I was thinking of populating the fields in the response message with bogus entries if that sounds like an acceptable error response?

tizianofiorenzani commented:
I think it's fair to just use the `type` field to check whether a parameter is valid.

Ryanf55 commented:
Overall, this is looking great. Do you know what additional services would be required to get this call to work?
```
$ ros2 param set /ardupilot_dds LOIT_SPEED 1249
Wait for service timed out
```

Ryanf55 commented on libraries/AP_DDS/AP_DDS_Client.cpp:
```diff
@@ -711,6 +718,194 @@ void AP_DDS_Client::on_request(uxrSession* uxr_session, uxrObjectId object_id, u
+        rcl_interfaces_srv_SetParameters_Response set_parameter_response;
+        const bool deserialize_success = rcl_interfaces_srv_SetParameters_Request_deserialize_topic(ub, &set_parameter_request);
+        if (deserialize_success == false) {
+            break;
```
It might be good to add a GSC_SEND_TEXT warning here. I don't think people will be calling this in a loop.

Ryanf55 commented on libraries/AP_DDS/AP_DDS_Client.cpp:
```diff
@@ -711,6 +718,194 @@ void AP_DDS_Client::on_request(uxrSession* uxr_session, uxrObjectId object_id, u
+            strncpy(param_key, (char *)param.name, AP_MAX_NAME_SIZE);
+            param_key[AP_MAX_NAME_SIZE] = 0;
+
+            // Only worried about integer and float types for parameter setting.
```
It could be useful to comment which param types we don't handle.

PQuill33 replied:
I added a note here in my latest commit for this.

Ryanf55 commented on libraries/AP_DDS/AP_DDS_Client.cpp:
```diff
@@ -711,6 +718,194 @@ void AP_DDS_Client::on_request(uxrSession* uxr_session, uxrObjectId object_id, u
+                break;
+            }
+            default: {
+                break;
```
Do we silently discard other parameter types? What's the expected user behavior in this case?

PQuill33 replied:
Yes, they get discarded, but the user will get a response message handled by setting `successful = false`  and the reason as "Parameter not found" 

If it is a non-supported parameter type, it will get caught in the if statement (lines 774-778) to set the successful flag and reason in the response message.

Ryanf55 commented on libraries/AP_DDS/AP_DDS_Client.cpp:
```diff
@@ -711,6 +718,194 @@ void AP_DDS_Client::on_request(uxrSession* uxr_session, uxrObjectId object_id, u
+
+            // Add existing parameter checks used in GCS_Param.cpp
+            if (parameter_flags & AP_PARAM_FLAG_INTERNAL_USE_ONLY) {
+                // the user can set BRD_OPTIONS to enable set of internal
+                // parameters, for developer testing or unusual use cases
```
```suggestion
                // The user can set BRD_OPTIONS to enable set of internal
                // parameters, for developer testing or unusual use cases.
```
Please end multi-line comments in periods which are easier to detect issues in merge conflicts.

Ryanf55 commented on libraries/AP_DDS/AP_DDS_Client.cpp:
```diff
@@ -711,6 +718,194 @@ void AP_DDS_Client::on_request(uxrSession* uxr_session, uxrObjectId object_id, u
+            vp->set_float(param_value, var_type);
+
+            // Force the save if the value is not equal to the old one
+            bool force_save = !is_equal(param_value, old_value);
```
Did you look at using `set_and_save_by_name_ifchanged`?

PQuill33 replied:
I did not, but after review it looks like it does the same thing. I updated the code to use this since it is more direct.

Ryanf55 commented on libraries/AP_DDS/AP_DDS_Client.cpp:
```diff
@@ -711,6 +718,194 @@ void AP_DDS_Client::on_request(uxrSession* uxr_session, uxrObjectId object_id, u
+
+            // set parameter
+            AP_Param *vp;
+            char param_key[AP_MAX_NAME_SIZE+1];
```
```suggestion
            char param_key[AP_MAX_NAME_SIZE + 1];
```

Style nit - `astyle` doesn't catch these.

Ryanf55 commented on libraries/AP_DDS/AP_DDS_Client.cpp:
```diff
@@ -711,6 +718,194 @@ void AP_DDS_Client::on_request(uxrSession* uxr_session, uxrObjectId object_id, u
+            enum ap_var_type var_type;
+
+            AP_Param *vp;
+            char param_key[AP_MAX_NAME_SIZE+1];
```
```suggestion
            char param_key[AP_MAX_NAME_SIZE + 1];
```

Style nit

Ryanf55 commented on libraries/AP_DDS/AP_DDS_Client.cpp:
```diff
@@ -711,6 +718,194 @@ void AP_DDS_Client::on_request(uxrSession* uxr_session, uxrObjectId object_id, u
+                continue;
+            }
+
+            float param_value = vp->cast_to_float(var_type);
```
```suggestion
            const float param_value = vp->cast_to_float(var_type);
```

Use const for locals when possible

Ryanf55 commented on libraries/AP_DDS/AP_DDS_Client.cpp:
```diff
@@ -711,6 +718,194 @@ void AP_DDS_Client::on_request(uxrSession* uxr_session, uxrObjectId object_id, u
+
+        uxr_buffer_reply(uxr_session, reliable_out, replier_id, sample_id, reply_buffer, ucdr_buffer_length(&reply_ub));
+
+        GCS_SEND_TEXT(MAV_SEVERITY_INFO, "%s Get Parameters Request : %s", msg_prefix, successful_read ? "SUCCESSFUL" : "ONE OR MORE PARAM NOT FOUND");
```
```suggestion
        GCS_SEND_TEXT(successful_read ? MAV_SEVERITY_INFO : MAV_SEVERITY_WARNING, "%s Get Parameters Request : %s", msg_prefix, successful_read ? "SUCCESSFUL" : "ONE OR MORE PARAM NOT FOUND");
```

I like this trick to make the log level different dependent on the result code.

PQuill33 replied:
Done, and I added the same trick in the set_parameters block.

Ryanf55 commented:
Needs a rebase on `master`.

PQuill33 commented:
I resolved the suggested changes and rebased onto master.

I used @Ryanf55 suggested change to get the ros2 cli to work with the parameter management, now it can also perform the following:
```
$ ros2 param set ap LOIT_SPEED 1350
Set parameter successful: Parameter accepted
$ ros2 param get ap LOIT_SPEED
Double value is: 1350.0
```

Ryanf55 commented on libraries/AP_DDS/AP_DDS_Client.cpp:
```diff
@@ -800,6 +808,188 @@ void AP_DDS_Client::on_request(uxrSession* uxr_session, uxrObjectId object_id, u
+            strncpy(param_key, (char *)get_parameters_request.names[i], AP_MAX_NAME_SIZE);
+            param_key[AP_MAX_NAME_SIZE] = 0;
+
+            vp = AP_Param::find(param_key, &var_type);
```
This could be a callback in the IO thread because this could take a few mS on hardware.

tridge commented on libraries/AP_DDS/AP_DDS_Client.cpp:
```diff
@@ -800,6 +808,188 @@ void AP_DDS_Client::on_request(uxrSession* uxr_session, uxrObjectId object_id, u
+            switch (var_type) {
+            case AP_PARAM_INT8:
+            case AP_PARAM_INT16:
+            case AP_PARAM_INT32: {
```
this is losing precision

PQuill33 replied:
Done in most recent commit

Ryanf55 commented on libraries/AP_DDS/AP_DDS_Client.cpp:
```diff
@@ -800,6 +808,188 @@ void AP_DDS_Client::on_request(uxrSession* uxr_session, uxrObjectId object_id, u
+                continue;
+            }
+
+            const float param_value = vp->cast_to_float(var_type);
```
We can get some higher precision if we don't always cast to float.

PQuill33 replied:
Updated in latest commit

tridge commented on libraries/AP_DDS/AP_DDS_Client.cpp:
```diff
@@ -800,6 +808,188 @@ void AP_DDS_Client::on_request(uxrSession* uxr_session, uxrObjectId object_id, u
+
+        uxr_buffer_reply(uxr_session, reliable_out, replier_id, sample_id, reply_buffer, ucdr_buffer_length(&reply_ub));
+
+        GCS_SEND_TEXT(successful_read ? MAV_SEVERITY_INFO : MAV_SEVERITY_WARNING, "%s Get Parameters Request : %s", msg_prefix, successful_read ? "SUCCESSFUL" : "ONE OR MORE PARAM NOT FOUND");
```
this is too verbose

PQuill33 replied:
I will address this in a follow up PR

Ryanf55 commented on libraries/AP_DDS/AP_DDS_Client.cpp:
```diff
@@ -800,6 +808,188 @@ void AP_DDS_Client::on_request(uxrSession* uxr_session, uxrObjectId object_id, u
+
+        uxr_buffer_reply(uxr_session, reliable_out, replier_id, sample_id, reply_buffer, ucdr_buffer_length(&reply_ub));
+
+        GCS_SEND_TEXT(successful_read ? MAV_SEVERITY_INFO : MAV_SEVERITY_WARNING, "%s Get Parameters Request : %s", msg_prefix, successful_read ? "SUCCESSFUL" : "ONE OR MORE PARAM NOT FOUND");
```
Let's wrap this in DDS debug statement within a define.

PQuill33 replied:
I will address this in a follow up PR

tridge reviewed (approved):
a followup PR for debug verbosity and precision of int32 vars would be nice

tridge commented:
needs fix for stack frame size on build

Ryanf55 commented:
> needs fix for stack frame size on build

```
../../libraries/AP_DDS/AP_DDS_Client.cpp: In member function 'void AP_DDS_Client::on_request(uxrSession*, uxrObjectId, uint16_t, SampleIdentity*, ucdrBuffer*, uint16_t)':
../../libraries/AP_DDS/AP_DDS_Client.cpp:995:1: error: the frame size of 2800 bytes is larger than 1300 bytes [-Werror=frame-larger-than=]
  995 | }
      | ^
compilation terminated due to -Wfatal-errors.
cc1plus: all warnings being treated as errors
```

I can help reproduce this if you need.

joshanne commented on libraries/AP_DDS/AP_DDS_Client.cpp:
```diff
@@ -800,6 +808,188 @@ void AP_DDS_Client::on_request(uxrSession* uxr_session, uxrObjectId object_id, u
+            // Check that all the parameters were set successfully
+            successful_params &= set_parameter_response.results[i].successful;
+        }
+        GCS_SEND_TEXT(successful_params ? MAV_SEVERITY_INFO : MAV_SEVERITY_WARNING, "%s Set Parameters Request : %s", msg_prefix, successful_params ? "SUCCESSFUL" : "ONE OR MORE PARAMS FAILED" );
```
A nice little pattern for compile time inclusion of debug prints:

https://github.com/ArduPilot/ardupilot/blob/de675641adcc8d4ec0cb4fc8d0bd587eded3ff04/libraries/AP_ONVIF/AP_ONVIF.cpp#L28

PQuill33 replied:
Thanks! This will help with the follow up PR to clean up the gcs messages and put them into debug statements.

Ryanf55 replied:
Are there any that support different log levels? Not everything we have in debug is info level.

PQuill33 commented:
> > needs fix for stack frame size on build
> 
> ```
> ../../libraries/AP_DDS/AP_DDS_Client.cpp: In member function 'void AP_DDS_Client::on_request(uxrSession*, uxrObjectId, uint16_t, SampleIdentity*, ucdrBuffer*, uint16_t)':
> ../../libraries/AP_DDS/AP_DDS_Client.cpp:995:1: error: the frame size of 2800 bytes is larger than 1300 bytes [-Werror=frame-larger-than=]
>   995 | }
>       | ^
> compilation terminated due to -Wfatal-errors.
> cc1plus: all warnings being treated as errors
> ```
> 
> I can help reproduce this if you need.

I've been able to reproduce it. The last commit I pushed brought the frame size down quite a bit. Just need to bring it down a little more.

I tried changing [this file]( https://github.com/PQuill33/ardupilot/blob/master/libraries/AP_DDS/wscript#L79)      with container_prealloc_size = 8    changed to 2 and it can build fine. I tried to build Durandal and pixracer with that change and it worked.

I also tried running the dds tests and nothing failed, but I wasn't sure if this is an appropriate path to take.

Ryanf55 commented:
> I tried changing [this file](https://github.com/PQuill33/ardupilot/blob/master/libraries/AP_DDS/wscript#L79) with container_prealloc_size = 8 changed to 2 and it can build fine. I tried to build Durandal and pixracer with that change and it worked.
> 
> I also tried running the dds tests and nothing failed, but I wasn't sure if this is an appropriate path to take.


Please leave that to 8, we need arrays larger than two for our dynamic transforms when doing SLAM. 
There's probably another place to increase this, perhaps here, on the `8192` value:
https://github.com/ArduPilot/ardupilot/blob/12b761c53499abc4514892fe626ee1af5ca2acd2/libraries/AP_DDS/AP_DDS_Client.cpp#L621

It's the stack size in this function:
https://github.com/ArduPilot/ardupilot/blob/e232ccde7d232588515e93f59a231c6b6d52fed8/libraries/AP_HAL/Scheduler.h#L124

PQuill33 commented:
I fixed the stack frame size issue on build and reworked the integer precision in the get_parameter service. All checks passed now.

Per the dev call yesterday, I plan to put together a smaller follow on PR to move the GCS_SEND_TEXT statements to a DDS debug statement and will address the verbosity in there as well.

Ryanf55 commented:
Need to add a wiki. Don't use params if you are doing realtime control.
